Conjur/conventions/ruby-style-guide.md

```diff
@@ -24,10 +23,194 @@ This guide is intended to be a "living document" in the sense that it's expected
+    audit_success
+    new_token
+  rescue => e
+    audit_failure(e)
+    raise e
+  end
+  ```
+
+  
+
+- Methods that return an object should have the name of the returned object so these functions should not have verb phrase names.
+   In some cases, these rules may conflict when a side-effectful method also returns a value. Here the "verb phrase" rule should win since indicating the side effect is more important.
+
+- For low-level style decisions about formatting, etc, we follow the ruby style. We advice to use rubocop to auto correct offence in the files you change before submitting to PR.
+
+- Avoid "name stutter". That is, don't repeat a name from the module path in a class, and don't repeat a name from the class in a method. For example:
+
+   ```ruby
+   # bad
+   class MyCar
+     def start_car_engine
+       # ...
+     end
+   end
+   
+   # good
+   class MyCar
+     def start_engine
+       # ...
+     end
+   end
+   
+   # bad
+   module Authenticators
+     class OktaAuthenticator
+       #...
+     end
+   end
+   
+   # good
+   module Authenticators
+     class Okta
+       #...
+     end
+   end
+   ```
+
+- In command class we should split methods called in `call` function into 2 parts - one with an explicit name so that the command class 'call' function tells the story, and one with the returned object name for memoization.
+
+  - Bad:
+
+    ```ruby
+    def get_jwt_identity_from_request
+        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingGetJwtIdentity.new)
+        @jwt_identity ||= @jwt_configuration.jwt_identity
+        @logger.info(LogMessages::Authentication::AuthnJwt::FoundJwtIdentity.new(jwt_identity))
+    end
+    ```
+
+  - Good:
+
+    ```ruby
+    def get_jwt_identity_from_request
+        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingGetJwtIdentity.new)
+        jwt_identity
+        @logger.info(LogMessages::Authentication::AuthnJwt::FoundJwtIdentity.new(jwt_identity))
+    end
+    
+    def jwt_identity
+        @jwt_identity ||= @jwt_configuration.jwt_identity
+    end
+    ```
+
+- In constuctor of classes separate input handling and dependency handling with empty line
+
+- In command class implement methods in the order they are called
+
+  ```ruby
+  def call
+    func_a
+    func_b
+    func_c
+  end
+  
+  private
+  
+  def func_a
+    ...
+  end
+  
+  def func_b
+    ...
+  end
+  
+  def sub_func_b
+    ...
+  end
+  
+  def func_c
+    ...
+    end
+  ```
+
+- In functions with many parameters put each parameter in seperate line in the declaration. In this case end each param with ":" so the caller will need to specify name of variable for each param.
+
+  ```
+  def method(
+    param1:
+    param2:
+    param3:
+  )
+  ```
+
+  
+
+- When overriding a code climate violation with an inline exception (this isn't possible for all code climate  plugins but is for many), add the reason for the exception on the line above the special "ignore" directive
```

Comment:
I'd expect this to be super uncommon. I've used overrides in the case of improved readability. The default complexity and method length settings can be problematic as well. Coding standards push people to write higher quality, easier to understand code. If there are a lot of warnings, it's more likely an issue of code not being as mature as we'd like it to be.

Reply:
Regarding this i think its the job of the PR reviewer to check each warning in the PR and decide if he accepts it or wants the code to change. WDYT

Reply:
I will remove this section and when we agree it will be done in another pr

Reply:
@jvanderhoof I don't take this rule as saying it's ok to do lots of overrides (though perhaps we should clarify that it _isn't_), but only "_If you _have to_ do an override for a legitimate reason, then you need to document the reason as a comment"